proposals/NNNN-this-address-space.md

```diff
@@ -0,0 +1,158 @@
+*   Status: **Design In Progress**
+
+*During the review process, add the following fields as needed:*
+
+*   PRs: [\#111](https://github.com/llvm/wg-hlsl/pull/111),
+    [llvm-project:\#127675](https://github.com/llvm/llvm-project/pull/127675)
+*   Issues:
+*   Posts:
+
+## Introduction
+
+HLSL uses copy-in/copy-out semantics for parameter passing. This requires that
+the pointer passed to the callee points to the default address space and
+directly to a variable. This simplifies code generation for the callee, as there
+is only one address space for the parameter. It also ensures that the SPIR-V
+restriction that pointer operands must be memory object declarations is true.
+
+However, this is not the case for the 'this' pointer on member functions. The
+HLSL specification states that the 'this' pointer is a reference to the object,
+which means that it could be a pointer to any address space and could point to
+an object that is a member of another object.
+
+## Motivation
+
+HLSL allows member functions. In HLSL, the this pointer is a reference to the
+object on which it was called. From the user perspective, the same function is
+called regardless of the address space of the object on which it is called. The
+user does not have to write a version of the function for each address space.
+
+This is implemented in DXC, and the same behavior must be implemented in Clang.
+So far, this has not been a problem because no address space other than the
+default address space has been used until recently. This problem will be exposed
+when the hlsl\_device and hlsl\_constant address spaces are used.
+
+Consider this example:
+
+```c
+struct S {
+  int a;
+  int add(int v) { return a+v; }
+};
+
+cbuffer B : register(b1) {
+  S s;
+};
+
+RWBuffer<int> o : register(u0);
+
+[numthreads(1,1,1)]
+void main()
+{
+  o[0] = s.add(3);
+}
+```
+
+Once
+[https://github.com/llvm/llvm-project/pull/124886](https://github.com/llvm/llvm-project/pull/124886),
+lands, this example fails with the error:
+
+```
+t.hlsl:15:10: error: cannot initialize object parameter of type 'S' with an expression of type 'hlsl_constant S'
+   15 |   o[0] = s.add(3);
+```
+
+## Proposed solution
+
+All HLSL address spaces will be made a subspace of a new hlsl\_generic address
+space. The \`this\` pointer will point to data within this address space. This
+allows an address space cast to be inserted on the actual parameter by Clang and
+the member function to be called.
+
+These address space casts cause issues for SPIR-V because SPIR-V’s Generic
+storage class is not allowed by the Vulkan environment for SPIR-V, and a pointer
+cannot be cast from one storage class to another. The address space casts will
+be removed through optimizations and a fix-up pass.
+
+All calls to member functions will be inlined, even if the function is marked as
+noinline. Once inlined, a pass that will propagate the address space before the
+cast to all uses of the address space cast can be run.
```

Comment:
maybe we should mention linking here:
 - if we do real libraries, then we'd need to keep the original function if exported, and apply the same transformation post-linking

Reply:
Since we don't have a good idea of how linking will work, I don't want to say what we will do in that case. I'm not sure how member functions will work. Are we allowed to mark member functions as export (https://godbolt.org/z/8WeEfrGGc)?